From 40cad8db1c685631cf873f46e8933aa74de98971 Mon Sep 17 00:00:00 2001
From: Valentin Caron <valentin.caron@foss.st.com>
Date: Mon, 2 Sep 2024 15:24:39 +0200
Subject: [PATCH 1229/1695] plat-stm32mp2: scmi_server: implement reset domain
 protocol

scmi_server_scpfw can now retrieve resets description from DT.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 2bf772aca0efd89b4176f8521e6255553c5cbf7e plat-stm32mp2: scmi_server: implement reset domain protocol]
Signed-off-by: Valentin Caron <valentin.caron@foss.st.com>
Change-Id: Ia20145c7cc7ee22c92b451d571d20cbccc415794
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/401863
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/arch/arm/plat-stm32mp2/conf.mk           |   2 +
 .../arm/plat-stm32mp2/scmi_server_scpfw.c     |   8 +
 .../scmi-server/include/scmi_reset_consumer.h |  34 +++++
 core/lib/scmi-server/scmi_reset_consumer.c    | 142 ++++++++++++++++++
 core/lib/scmi-server/sub.mk                   |   1 +
 5 files changed, 187 insertions(+)
 create mode 100644 core/lib/scmi-server/include/scmi_reset_consumer.h
 create mode 100644 core/lib/scmi-server/scmi_reset_consumer.c

diff --git a/core/arch/arm/plat-stm32mp2/conf.mk b/core/arch/arm/plat-stm32mp2/conf.mk
index 5675a1056..7e1afa69e 100644
--- a/core/arch/arm/plat-stm32mp2/conf.mk
+++ b/core/arch/arm/plat-stm32mp2/conf.mk
@@ -93,6 +93,7 @@ $(call force,CFG_SCMI_SCPFW,n)
 $(call force,CFG_SCMI_SERVER_CLOCK_CONSUMER,n)
 $(call force,CFG_SCMI_SERVER_PD_CONSUMER,n)
 $(call force,CFG_SCMI_SERVER_REGULATOR_CONSUMER,n)
+$(call force,CFG_SCMI_SERVER_RESET_CONSUMER,n)
 $(call force,CFG_STM32_BSEC_PTA,n)
 $(call force,CFG_STM32_BSEC3,n)
 $(call force,CFG_STM32_BSEC_WRITE,n)
@@ -196,6 +197,7 @@ CFG_SCMI_SCPFW ?= y
 CFG_SCMI_SERVER_CLOCK_CONSUMER ?= y
 CFG_SCMI_SERVER_PD_CONSUMER ?= y
 CFG_SCMI_SERVER_REGULATOR_CONSUMER ?= y
+CFG_SCMI_SERVER_RESET_CONSUMER ?= y
 
 CFG_DRIVERS_I2C ?= $(CFG_STM32_I2C)
 CFG_REGULATOR_FIXED ?= y
diff --git a/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c b/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c
index f4058aaf1..de9e0c05e 100644
--- a/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c
+++ b/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c
@@ -18,6 +18,7 @@
 #include <scmi_clock_consumer.h>
 #include <scmi_pd_consumer.h>
 #include <scmi_regulator_consumer.h>
+#include <scmi_reset_consumer.h>
 #include <scmi/scmi_server.h>
 #include <tee_api_types.h>
 #include <tee_api_defines.h>
@@ -428,6 +429,13 @@ static TEE_Result optee_scmi_server_probe(const void *fdt, int parent_node,
 				if (res)
 					panic("Error during regulators init");
 				break;
+			case SCMI_PROTOCOL_ID_RESET_DOMAIN:
+				res = optee_scmi_server_init_resets(fdt,
+					protocol_ctx->dt_node, agent_cfg,
+					channel_cfg);
+				if (res)
+					panic("Error during resets init");
+				break;
 			default:
 				EMSG("%s Unknown protocol ID: %#x",
 				     protocol_ctx->dt_name,
diff --git a/core/lib/scmi-server/include/scmi_reset_consumer.h b/core/lib/scmi-server/include/scmi_reset_consumer.h
new file mode 100644
index 000000000..225d0bfea
--- /dev/null
+++ b/core/lib/scmi-server/include/scmi_reset_consumer.h
@@ -0,0 +1,34 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2024, STMicroelectronics
+ */
+
+#ifndef SCMI_SERVER_RESET_CONSUMER_H
+#define SCMI_SERVER_RESET_CONSUMER_H
+
+#include <scmi_agent_configuration.h>
+#include <stdint.h>
+#include <tee_api_types.h>
+
+#ifdef CFG_SCMI_SERVER_RESET_CONSUMER
+/*
+ * Initialize SCMI reset from reset consumer information from DT.
+ * Returns a TEE_Result compliant value
+ */
+TEE_Result optee_scmi_server_init_resets(const void *fdt, int node,
+					 struct scpfw_agent_config *agent_cfg,
+					 struct scpfw_channel_config
+							*channel_cfg);
+#else
+static inline TEE_Result
+	optee_scmi_server_init_resets(const void *fdt __unused,
+				      int node __unused,
+				      struct scpfw_agent_config *agent_cfg
+						__unused,
+				      struct scpfw_channel_config *channel_cfg
+						__unused)
+{
+	return TEE_SUCCESS;
+}
+#endif /* CFG_SCMI_SERVER_RESET_CONSUMER */
+#endif /* SCMI_SERVER_RESET_CONSUMER_H */
diff --git a/core/lib/scmi-server/scmi_reset_consumer.c b/core/lib/scmi-server/scmi_reset_consumer.c
new file mode 100644
index 000000000..c3251ae53
--- /dev/null
+++ b/core/lib/scmi-server/scmi_reset_consumer.c
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2024, STMicroelectronics
+ */
+
+#include <assert.h>
+#include <drivers/rstctrl.h>
+#include <kernel/dt.h>
+#include <kernel/panic.h>
+#include <libfdt.h>
+#include <malloc.h>
+#include <scmi_agent_configuration.h>
+#include <scmi_reset_consumer.h>
+#include <tee_api_defines_extensions.h>
+#include <trace.h>
+
+/*
+ * struct scmi_server_reset: data for a SCMI reset in DT
+ *
+ * @domain_id: SCMI domain identifier
+ * @domain_name: SCMI domain name
+ * @reset: reset to control thru SCMI protocol
+ */
+struct scmi_server_reset {
+	uint32_t domain_id;
+	const char *domain_name;
+	struct rstctrl *reset;
+};
+
+TEE_Result optee_scmi_server_init_resets(const void *fdt, int node,
+					 struct scpfw_agent_config *agent_cfg,
+					 struct scpfw_channel_config
+							*channel_cfg)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct scmi_server_reset *s_resets = NULL;
+	size_t s_resets_count = 0;
+	int subnode = 0, item_node = 0;
+	bool have_subnodes = false;
+	size_t n = 0;
+
+	item_node = fdt_subnode_offset(fdt, node, "resets");
+	if (item_node < 0)
+		return TEE_SUCCESS;
+
+	/* Compute the number of domains to allocate */
+	fdt_for_each_subnode(subnode, fdt, item_node) {
+		paddr_t reg = fdt_reg_base_address(fdt, subnode);
+
+		assert(reg != DT_INFO_INVALID_REG);
+		if ((size_t)reg > s_resets_count)
+			s_resets_count = (uint32_t)reg;
+
+		have_subnodes = true;
+	}
+
+	if (!have_subnodes)
+		return TEE_SUCCESS;
+
+	s_resets_count += 1;
+	s_resets = calloc(s_resets_count, sizeof(*s_resets));
+	if (!s_resets)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	fdt_for_each_subnode(subnode, fdt, item_node) {
+		struct scmi_server_reset *s_reset = NULL;
+		struct rstctrl *reset = NULL;
+		const fdt32_t *cuint = NULL;
+		uint32_t domain_id = 0;
+
+		res = rstctrl_dt_get_by_index(fdt, subnode, 0, &reset);
+		if (res == TEE_ERROR_DEFER_DRIVER_INIT) {
+			panic("Unexpected init deferral");
+		} else if (res) {
+			EMSG("Can't get reset %s (%#"PRIx32"), skipped",
+			     fdt_get_name(fdt, subnode, NULL), res);
+			continue;
+		}
+
+		domain_id = (uint32_t)fdt_reg_base_address(fdt, subnode);
+		s_reset = s_resets + domain_id;
+		s_reset->domain_id = domain_id;
+
+		cuint = fdt_getprop(fdt, subnode, "domain-name", NULL);
+		if (cuint)
+			s_reset->domain_name = (char *)cuint;
+		else
+			s_reset->domain_name = fdt_get_name(fdt, subnode, NULL);
+
+		s_reset->domain_name = strdup(s_reset->domain_name);
+		if (!s_reset->domain_name) {
+			free(s_resets);
+			return TEE_ERROR_OUT_OF_MEMORY;
+		}
+
+		/* Check that the domain_id is not already used */
+		if (s_reset->reset) {
+			EMSG("Domain ID %"PRIu32" already used", domain_id);
+			panic();
+		}
+		s_reset->reset = reset;
+
+		DMSG("scmi reset shares %s on domain ID %"PRIu32,
+		     s_reset->domain_name, domain_id);
+	}
+
+	/*
+	 * Assign domain IDs to un-exposed reset as SCMI specification
+	 * require the resource is defined even if not accessible.
+	 */
+	for (n = 0; n < s_resets_count; n++) {
+		if (!s_resets[n].reset) {
+			s_resets[n].domain_id = n;
+			s_resets[n].domain_name = "";
+		}
+	}
+
+	if (channel_cfg->reset) {
+		EMSG("Reset already loaded: agent %u, channel %u",
+		     agent_cfg->agent_id, channel_cfg->channel_id);
+		panic();
+	}
+
+	channel_cfg->reset_count = s_resets_count;
+	channel_cfg->reset = calloc(channel_cfg->reset_count,
+				    sizeof(*channel_cfg->reset));
+	if (!channel_cfg->reset) {
+		free(s_resets);
+		return TEE_ERROR_OUT_OF_MEMORY;
+	}
+
+	for (n = 0; n < s_resets_count; n++) {
+		unsigned int domain_id = s_resets[n].domain_id;
+
+		channel_cfg->reset[domain_id] = (struct scmi_reset){
+			.name = s_resets[n].domain_name,
+			.rstctrl = s_resets[n].reset,
+		};
+	}
+
+	return TEE_SUCCESS;
+}
diff --git a/core/lib/scmi-server/sub.mk b/core/lib/scmi-server/sub.mk
index bfbc8bc0e..4d4637b3c 100644
--- a/core/lib/scmi-server/sub.mk
+++ b/core/lib/scmi-server/sub.mk
@@ -29,6 +29,7 @@ srcs-y += scmi_server.c
 srcs-$(CFG_SCMI_SERVER_CLOCK_CONSUMER) += scmi_clock_consumer.c
 srcs-$(CFG_SCMI_SERVER_PD_CONSUMER) += scmi_pd_consumer.c
 srcs-$(CFG_SCMI_SERVER_REGULATOR_CONSUMER) += scmi_regulator_consumer.c
+srcs-$(CFG_SCMI_SERVER_RESET_CONSUMER) += scmi_reset_consumer.c
 global-incdirs-y += include
 
 # SCP-firmware cmake configuration generates header fwk_module_idx.h and
-- 
2.47.2

